cli/src/pcluster/models/imagebuilder.py

```diff
@@ -368,6 +387,52 @@ def _upload_artifacts(self):
                 f"Unable to upload imagebuilder cfn template to the S3 bucket {self.bucket.name} due to exception: {e}"
             )
 
+    def delete(self, force=False):
+        """Delete CFN Stack and associate resources and deregister the image."""
+        # validate image name
+        self._validate_image_name()
+
+        # TODO: check image is shared
+        if force or not self._check_instance_using_image():
+            try:
+                if AWSApi.instance().ec2.image_exists(self.image_name):
+                    # Deregister image
+                    AWSApi.instance().ec2.deregister_image(self.image.id)
+
+                    # Delete snapshot
+                    for snapshot_id in self.image.snapshot_ids:
+                        AWSApi.instance().ec2.delete_snapshot(snapshot_id)
+                elif AWSApi.instance().cfn.stack_exists(self.image_name):
+                    # Delete stack
+                    AWSApi.instance().cfn.delete_stack(self.image_name)
+
+                # Delete s3 image directory
+                try:
+                    self.bucket.check_bucket_exists()
+                    self.bucket.delete_s3_artifacts()
+                except AWSClientError:
+                    logging.warning("S3 bucket %s does not exist, skip image s3 artifacts deletion.", self.bucket.name)
+
+            except AWSClientError as e:
+                raise ImageBuilderActionError(f"Unable to delete image and stack, due to {str(e)}")
```

Comment:
Let's double think to error cases, e.g. you don't have permission to deregister_image but you have permissions to delete_snapshots or viceversa. Is it correct that we're failing at the first action, or should we proceed with the actions we can do?

Reply:
It depends on what kind of user experience we want to deliver.
case 1. A boto3 client operation failed due to  permission, we raise the error and let user manually add the permission and run command again.
case 2. A boto3 client operation failed, we log a warning message and continue executing the command.

For now, we are developing on case 1, raise error when there is a boto3 client operation failure.

Reply:
In this case snapshots cannot be deleted if image is not deregistered, so it's ok to fail fast